components/layout_2020/taffy/layout.rs

```diff
@@ -0,0 +1,620 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+use std::marker::PhantomData;
+
+use app_units::Au;
+use atomic_refcell::{AtomicRef, AtomicRefCell, AtomicRefMut};
+use style::properties::ComputedValues;
+use style::values::generics::length::{GenericLengthPercentageOrAuto, LengthPercentageOrAuto};
+use style::values::specified::align::AlignFlags;
+use style::values::specified::box_::DisplayInside;
+use style::Zero;
+use taffy::style_helpers::{TaffyMaxContent, TaffyMinContent};
+use taffy::MaybeMath;
+
+use super::{TaffyContainer, TaffyItemBox, TaffyItemBoxInner, TaffyStyloStyle};
+use crate::cell::ArcRefCell;
+use crate::context::LayoutContext;
+use crate::formatting_contexts::{Baselines, IndependentFormattingContext, IndependentLayout};
+use crate::fragment_tree::{BoxFragment, CollapsedBlockMargins, Fragment};
+use crate::geom::{
+    AuOrAuto, LogicalSides, LogicalVec2, PhysicalPoint, PhysicalRect, PhysicalSides, PhysicalSize,
+    Size,
+};
+use crate::positioned::{AbsolutelyPositionedBox, PositioningContext, PositioningContextLength};
+use crate::sizing::{ContentSizes, InlineContentSizesResult};
+use crate::style_ext::ComputedValuesExt;
+use crate::{ContainingBlock, IndefiniteContainingBlock};
+
+const DUMMY_NODE_ID: taffy::NodeId = taffy::NodeId::new(u64::MAX);
+
+fn resolve_content_size(constraint: taffy::AvailableSpace, content_sizes: ContentSizes) -> f32 {
+    match constraint {
+        taffy::AvailableSpace::Definite(limit) => {
+            let min = content_sizes.min_content.to_f32_px();
+            let max = content_sizes.max_content.to_f32_px();
+            limit.min(max).max(min)
+        },
+        taffy::AvailableSpace::MinContent => content_sizes.min_content.to_f32_px(),
+        taffy::AvailableSpace::MaxContent => content_sizes.max_content.to_f32_px(),
+    }
+}
+
+#[inline(always)]
+fn with_independant_formatting_context<T>(
+    item: &mut TaffyItemBoxInner,
+    cb: impl FnOnce(&mut IndependentFormattingContext) -> T,
+) -> T {
+    match item {
+        TaffyItemBoxInner::InFlowBox(ref mut context) => cb(context),
+        TaffyItemBoxInner::OutOfFlowAbsolutelyPositionedBox(ref abspos_box) => {
+            let mut abspos_box = AtomicRefCell::borrow_mut(abspos_box);
+            cb(&mut abspos_box.context)
+        },
+    }
+}
+
+/// Layout parameters and intermediate results about a flex container,
+/// grouped to avoid passing around many parameters
+struct FlexContext<'a> {
+    source_child_nodes: &'a [ArcRefCell<TaffyItemBox>],
+    layout_context: &'a LayoutContext<'a>,
+    positioning_context: &'a mut PositioningContext,
+    content_box_size_override: &'a ContainingBlock<'a>,
+    style: &'a ComputedValues,
+}
```

Comment:
```suggestion
/// Layout parameters and intermediate results about a grid container,
/// grouped to avoid passing around many parameters
struct GridContext<'a> {
    source_child_nodes: &'a [ArcRefCell<TaffyItemBox>],
    layout_context: &'a LayoutContext<'a>,
    positioning_context: &'a mut PositioningContext,
    content_box_size_override: &'a ContainingBlock<'a>,
    style: &'a ComputedValues,
}
```

Reply:
I've gone with `TaffyContainerContext` which matches the "taffy" naming convention of the rest of module, and reflects that almost none of the code is grid-specific (only the implementation of `taffy::LayoutGridContainer` and the calls to `taffy::compute_grid_layout`)